dangerzone/isolation_provider.py

```diff
@@ -0,0 +1,501 @@
+            except:
+                log.warning("Couldn't delete old container image, so leaving it there")
+
+        return installed
+
+    def _parse_progress(self, document: Document, line: str) -> Tuple[bool, str, int]:
+        """
+        Parses a line returned by the container.
+        """
+        try:
+            status = json.loads(line)
+        except:
+            error_message = f"Invalid JSON returned from container:\n\n\t {line}"
+            log.error(error_message)
+            return (True, error_message, -1)
+
+        s = Style.BRIGHT + Fore.YELLOW + f"[doc {document.id}] "
+        s += Fore.CYAN + f"{status['percentage']}% "
+        if status["error"]:
+            s += Style.RESET_ALL + Fore.RED + status["text"]
+            log.error(s)
+        else:
+            s += Style.RESET_ALL + status["text"]
+            log.info(s)
+
+        return (status["error"], status["text"], status["percentage"])
+
+    def _exec(
+        self,
+        document: Document,
+        args: List[str],
+        stdout_callback: Optional[Callable] = None,
+    ) -> int:
+        args_str = " ".join(pipes.quote(s) for s in args)
+        log.info("> " + args_str)
+
+        with subprocess.Popen(
+            args,
+            stdin=None,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            bufsize=1,
+            universal_newlines=True,
+            startupinfo=startupinfo,
+        ) as p:
+            if p.stdout is not None:
+                for line in p.stdout:
+                    (error, text, percentage) = self._parse_progress(document, line)
+                    if stdout_callback:
+                        stdout_callback(error, text, percentage)
+
+            p.communicate()
+            return p.returncode
+
+    def _exec_container(
+        self,
+        document: Document,
+        command: List[str],
+        extra_args: List[str] = [],
+        stdout_callback: Optional[Callable] = None,
+    ) -> int:
+        container_runtime = self.get_runtime()
+
+        if self._get_runtime_name() == "podman":
+            platform_args = []
+            security_args = ["--security-opt", "no-new-privileges"]
+            security_args += ["--userns", "keep-id"]
+        else:
+            platform_args = ["--platform", "linux/amd64"]
+            security_args = ["--security-opt=no-new-privileges:true"]
+
+        # drop all linux kernel capabilities
+        security_args += ["--cap-drop", "all"]
+        user_args = ["-u", "dangerzone"]
+
+        prevent_leakage_args = ["--rm"]
+
+        args = (
+            ["run", "--network", "none"]
+            + platform_args
+            + user_args
+            + security_args
+            + prevent_leakage_args
+            + extra_args
+            + [self.CONTAINER_NAME]
+            + command
+        )
+
+        args = [container_runtime] + args
+        return self._exec(document, args, stdout_callback)
+
+    def _convert(
```

Comment:
Hot take: everything aside from the command creation and execution should be factored out of this class, so that the `Dummy` provider can use it as well. This will improve the testing surface a lot, as we will be able to set a GUI option (e.g., change the prefix), and see it actually executed.

Reply:
Why would the following make sense in `Dummy`? `is_container_installed()`, `get_runtime_name()` and `get_runtime()`. Theoretically, wouldn't we have tests for each isolation provider? In that sense we could test coverage specifically for `Container`.

Other than that, I think `parse_progress()` could make sense to get factored out.

Reply:
Actually, I'm referring to the code in the `_convert()` method, not the methods outside of it.

Reply:
I agree. I didn't do it originally because passing the extra variables to `_convert()` would make it look like a train of arguments. But actually it's just two extra. So let's do that. 

Let me know what you think: [cb325ca](https://github.com/freedomofpress/dangerzone/pull/302/commits/cb325ca386d2c12f6872a319a27ca1a06ef986b6).

Reply:
It's in a right direction, but I'd like to press a bit more. If we also factor out the last part, where we move the converted file to the safe dir, then we can improve our test coverage as well (see this [thread](https://github.com/freedomofpress/dangerzone/pull/303#discussion_r1071362787)).

Reply:
Oh. Right, I mean to include, but forgot about it. Thanks!

Reply:
Now thinking a bit more about this, I'm not sure it makes much sense.

I think it's clearer if the `_convert()` does the conversion, returns a success bool and saves the final file in `document.output_filename`. How it does it should be implementation-specific: it may create a `/safe`, for example and put the intermediate file there, but all that matters is that in the end if `success is True`, then the output file must be in `document.output_filename` and we can test if that happened.

This extra flexibility would allow a future implementation (e.g. Qubes) where it reads directly from the `pixels_dir` and converts to the output PDF without needing the `safe_dir`. It could move the file directly to `document.output_filename`. 

Does this make sense? 

> [..] as we will be able to set a GUI option (e.g., change the prefix), and see it actually executed.

This thing, for example, won't be a problem since if there is a prefix change, that will be reflected in `document.output_filename` and the conversion will know to move the final file there.

So I propose we leave it as it originally was. Thoughts?

Reply:
Per verbal discussion, lets leave this in its original form.

Reply:
I have now revered the changes and save a dummy PDF so all tests pass on platforms where the dummy converter will be enabled.